File: python/mq-bridge-py/analysis/bench_http_native.py
"""Native-load HTTP benchmark for mq-bridge vs other Python frameworks.

This harness drives each server with an external, non-Python load generator
(``wrk``) instead of an in-process ``http.client`` loop. An in-process Python
client is GIL-bound and plateaus around ~13k req/s regardless of the server
under test, which makes every framework look the same and hides mq-bridge's
real throughput. ``wrk`` removes that client-side ceiling so the numbers
reflect the server, not the load tool.

Usage:
    uv run python analysis/bench_http_native.py --connections 1,8,32
    uv run python analysis/bench_http_native.py --targets mqb-worker,mqb-direct,faststream

Requires ``wrk`` on PATH (``brew install wrk``). The FastAPI/Starlette/Sanic/
aiohttp/FastStream targets additionally require those packages to be importable.
"""

from __future__ import annotations

import argparse
import os
import re
import shutil
import socket
import subprocess
import sys
import tempfile
import textwrap
import time
from contextlib import contextmanager
from pathlib import Path

HOST = "127.0.0.1"
PATH = "/bench"
KIND = "bench.tick"
BODY = '{"value":0}'

REQ_RE = re.compile(r"Requests/sec:\s*([0-9.]+)")
LAT_RE = re.compile(r"Latency\s+([0-9.]+\w+)")


def parse_csv_ints(value: str) -> list[int]:
    out = [int(p.strip()) for p in value.split(",") if p.strip()]
    if not out or any(v < 1 for v in out):
        raise argparse.ArgumentTypeError("expected positive integers")
    return out


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--connections", type=parse_csv_ints, default=[1, 8, 32])
    parser.add_argument("--duration", type=int, default=5, help="wrk duration seconds")
    parser.add_argument(
        "--targets",
        default="mqb-worker,mqb-direct,faststream,fastapi,starlette,sanic,aiohttp",
        help=(
            "Comma-separated subset of: mqb-worker, mqb-direct, faststream, "
            "fastapi, starlette, sanic, aiohttp."
        ),
    )
    parser.add_argument("--route-concurrency", type=int, default=8)
    return parser.parse_args()


def free_port() -> int:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.bind((HOST, 0))
        return int(sock.getsockname()[1])


def wait_for_port(port: int, timeout: float = 15.0) -> None:
    deadline = time.perf_counter() + timeout
    while time.perf_counter() < deadline:
        try:
            with socket.create_connection((HOST, port), timeout=0.2):
                return
        except OSError:
            time.sleep(0.02)
    raise TimeoutError(f"server on {HOST}:{port} did not become ready")


def wrk_lua(send_kind: bool) -> str:
    lines = [
        'wrk.method = "POST"',
        f"wrk.body = '{BODY}'",
        'wrk.headers["Content-Type"] = "application/json"',
    ]
    if send_kind:
        lines.append(f'wrk.headers["kind"] = "{KIND}"')
    return "\n".join(lines) + "\n"


def run_wrk(port: int, connections: int, duration: int, lua_path: str) -> tuple[float, str]:
    threads = min(connections, os.cpu_count() or 8)
    out = subprocess.run(
        [
            "wrk",
            f"-t{threads}",
            f"-c{connections}",
            f"-d{duration}s",
            "-s",
            lua_path,
            f"http://{HOST}:{port}{PATH}",
        ],
        check=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
    ).stdout
    req = REQ_RE.search(out)
    lat = LAT_RE.search(out)
    return (float(req.group(1)) if req else 0.0, lat.group(1) if lat else "?")


@contextmanager
def mqb_server(executor: str, route_concurrency: int):
    port = free_port()
    cfg = textwrap.dedent(
        f"""
        routes:
          http_bench:
            concurrency: {route_concurrency}
            batch_size: 128
            input:
              http:
                url: "{HOST}:{port}"
                path: "{PATH}"
                method: "POST"
                internal_buffer_size: 8192
                request_timeout_ms: 30000
                concurrency_limit: 512
                inline_response_fast_path: true
            output:
              response: {{}}
        """
    ).lstrip()
    launcher = textwrap.dedent(
        f"""
        from mq_bridge import Route
        route = Route.from_yaml({_repr(cfg)}, "http_bench")
        route.add_handler("{KIND}", lambda data: {{"value": data["value"] + 1}})
        route.run()
        """
    ).lstrip()
    env = os.environ.copy()
    if executor == "direct":
        env["MQ_BRIDGE_PY_HANDLER_EXECUTOR"] = "direct"
    else:
        env.pop("MQ_BRIDGE_PY_HANDLER_EXECUTOR", None)
    proc = subprocess.Popen([sys.executable, "-c", launcher], env=env)
    try:
        wait_for_port(port)
        yield port, True  # send_kind=True
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()


def _repr(text: str) -> str:
    # Embed the YAML as a Python literal; Route.from_yaml expects a path OR text
    # depending on version, so write to a temp file and pass the path instead.
    path = Path(tempfile.mkdtemp(prefix="mqb-native-")) / "route.yaml"
    path.write_text(text, encoding="utf-8")
    return repr(str(path))


@contextmanager
def faststream_server():
    port = free_port()
    app = textwrap.dedent(
        """
        import json
        from faststream.asgi import AsgiFastStream

        async def bench(scope, receive, send):
            if scope["type"] != "http" or scope["method"] != "POST":
                await send({"type": "http.response.start", "status": 404, "headers": []})
                await send({"type": "http.response.body", "body": b""})
                return
            body = bytearray(); more = True
            while more:
                e = await receive(); body.extend(e.get("body", b"")); more = e.get("more_body", False)
            data = json.loads(body); data["value"] += 1
            rb = json.dumps(data, separators=(",", ":")).encode()
            await send({"type": "http.response.start", "status": 200,
                        "headers": [(b"content-type", b"application/json")]})
            await send({"type": "http.response.body", "body": rb})

        app = AsgiFastStream(asgi_routes=[("/bench", bench)], logger=None)
        """
    ).lstrip()
    tmp = Path(tempfile.mkdtemp(prefix="mqb-fs-"))
    (tmp / "fs_app.py").write_text(app, encoding="utf-8")
    proc = subprocess.Popen(
        [sys.executable, "-m", "uvicorn", "fs_app:app", "--host", HOST,
         "--port", str(port), "--log-level", "warning"],
        cwd=str(tmp),
        # Suppress the cosmetic lifespan-teardown traceback uvicorn/FastStream
        # prints when terminated mid-run; it does not affect the measurement.
        stderr=subprocess.DEVNULL,
    )
    try:
        wait_for_port(port)
        yield port, False  # send_kind=False
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()


@contextmanager
def _spawn_python(source: str, prefix: str, port: int):
    """Write ``source`` to a temp module and run it as its own process."""
    tmp = Path(tempfile.mkdtemp(prefix=prefix))
    script = tmp / "server.py"
    script.write_text(textwrap.dedent(source).lstrip(), encoding="utf-8")
    proc = subprocess.Popen(
        [sys.executable, str(script)],
        cwd=str(tmp),
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    try:
        wait_for_port(port)
        yield port, False  # send_kind=False (only mq-bridge routes on the kind header)
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()


@contextmanager
def _uvicorn_asgi(app_source: str, prefix: str):
    """Run an ASGI app (exposed as ``app``) under uvicorn, one worker."""
    port = free_port()
    tmp = Path(tempfile.mkdtemp(prefix=prefix))
    (tmp / "app_mod.py").write_text(textwrap.dedent(app_source).lstrip(), encoding="utf-8")
    proc = subprocess.Popen(
        [sys.executable, "-m", "uvicorn", "app_mod:app", "--host", HOST,
         "--port", str(port), "--log-level", "warning"],
        cwd=str(tmp),
        stderr=subprocess.DEVNULL,
    )
    try:
        wait_for_port(port)
        yield port, False
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()


def fastapi_server():
    return _uvicorn_asgi(
        f"""
        from fastapi import FastAPI, Request
        from fastapi.responses import JSONResponse

        app = FastAPI()

        @app.post("{PATH}")
        async def bench(req: Request):
            data = await req.json()
            data["value"] += 1
            return JSONResponse(data)
        """,
        prefix="mqb-fastapi-",
    )


def starlette_server():
    return _uvicorn_asgi(
        f"""
        from starlette.applications import Starlette
        from starlette.responses import JSONResponse
        from starlette.routing import Route

        async def bench(request):
            data = await request.json()
            data["value"] += 1
            return JSONResponse(data)

        app = Starlette(routes=[Route("{PATH}", bench, methods=["POST"])])
        """,
        prefix="mqb-starlette-",
    )


def sanic_server():
    port = free_port()
    return _spawn_python(
        f"""
        from sanic import Sanic
        from sanic.response import json as sjson

        app = Sanic("bench")

        @app.post("{PATH}")
        async def bench(request):
            data = request.json
            data["value"] += 1
            return sjson(data)

        if __name__ == "__main__":
            app.run(host="{HOST}", port={port}, access_log=False,
                    single_process=True, motd=False)
        """,
        prefix="mqb-sanic-",
        port=port,
    )


def aiohttp_server():
    port = free_port()
    return _spawn_python(
        f"""
        from aiohttp import web

        async def bench(request):
            data = await request.json()
            data["value"] += 1
            return web.json_response(data)

        app = web.Application()
        app.add_routes([web.post("{PATH}", bench)])
        web.run_app(app, host="{HOST}", port={port}, print=None, access_log=None)
        """,
        prefix="mqb-aiohttp-",
        port=port,
    )


TARGETS = {
    "mqb-worker": lambda a: mqb_server("worker", a.route_concurrency),
    "mqb-direct": lambda a: mqb_server("direct", a.route_concurrency),
    "faststream": lambda a: faststream_server(),
    "fastapi": lambda a: fastapi_server(),
    "starlette": lambda a: starlette_server(),
    "sanic": lambda a: sanic_server(),
    "aiohttp": lambda a: aiohttp_server(),
}


def main() -> None:
    if shutil.which("wrk") is None:
        raise SystemExit("wrk not found on PATH (brew install wrk)")
    args = parse_args()
    targets = [t.strip() for t in args.targets.split(",") if t.strip()]
    for t in targets:
        if t not in TARGETS:
            raise SystemExit(f"unknown target '{t}'; choices: {', '.join(TARGETS)}")

    results: dict[str, dict[int, float]] = {}
    for target in targets:
        results[target] = {}
        with TARGETS[target](args) as (port, send_kind):
            lua = Path(tempfile.mkdtemp(prefix="mqb-wrk-")) / "post.lua"
            lua.write_text(wrk_lua(send_kind), encoding="utf-8")
            # warmup
            run_wrk(port, max(args.connections), 2, str(lua))
            for conn in args.connections:
                rps, lat = run_wrk(port, conn, args.duration, str(lua))
                results[target][conn] = rps
                print(f"{target:14s} c={conn:<4d} {rps:12,.0f} req/s  (lat {lat})", flush=True)

    print("\nreq/s by connections (native wrk load):")
    header = "target".ljust(14) + "".join(f"{c:>14d}" for c in args.connections)
    print(header)
    for target in targets:
        row = target.ljust(14) + "".join(
            f"{results[target].get(c, 0):>14,.0f}" for c in args.connections
        )
        print(row)


if __name__ == "__main__":
    main()

File: node/mq-bridge-node/analysis/bench_http_native.js
"use strict";

/**
 * Native-load HTTP benchmark for the mq-bridge Node binding vs other Node
 * HTTP frameworks.
 *
 * Like the Python `analysis/bench_http_native.py`, this drives each server with
 * an external load generator (`wrk`) rather than an in-process client loop, so
 * the numbers reflect the server and not the load tool. Each server runs in its
 * own child process; `wrk` posts a tiny JSON body and the server echoes it with
 * `value` incremented.
 *
 * Usage:
 *   node analysis/bench_http_native.js --connections 1,8,32 --duration 8
 *   node analysis/bench_http_native.js --targets mqb,node-http,fastify
 *
 * Requires `wrk` on PATH (`brew install wrk`). The `fastify` / `express`
 * targets additionally require those packages to be importable (install with
 * `npm i -D fastify express` in node/mq-bridge-node). The `mqb` target needs
 * the addon built with the `http` feature (`npm run build` or `npm run
 * build:ci`).
 */

const net = require("node:net");
const path = require("node:path");
const os = require("node:os");
const fs = require("node:fs");
const { spawn, spawnSync } = require("node:child_process");

const HOST = "127.0.0.1";
const PATH = "/bench";
const KIND = "bench.tick";
const BODY = '{"value":0}';

const REQ_RE = /Requests\/sec:\s*([0-9.]+)/;
const LAT_RE = /Latency\s+([0-9.]+\w+)/;

// Absolute path to this package's entry point, so child processes can require
// the binding regardless of their cwd.
const MQB_MAIN = require.resolve("..");

function parseCsvInts(value) {
  const out = value
    .split(",")
    .map((p) => p.trim())
    .filter(Boolean)
    .map((p) => Number.parseInt(p, 10));
  if (!out.length || out.some((v) => !Number.isInteger(v) || v < 1)) {
    throw new Error(`expected positive integers, got '${value}'`);
  }
  return out;
}

function parseArgs(argv) {
  const args = {
    connections: [1, 8, 32],
    duration: 5,
    targets: "mqb,node-http,uws,fastify,express",
    routeConcurrency: 8,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = () => argv[(i += 1)];
    if (arg === "--connections") args.connections = parseCsvInts(next());
    else if (arg === "--duration") args.duration = Number.parseInt(next(), 10);
    else if (arg === "--targets") args.targets = next();
    else if (arg === "--route-concurrency") args.routeConcurrency = Number.parseInt(next(), 10);
    else throw new Error(`unknown argument '${arg}'`);
  }
  if (!Number.isInteger(args.duration) || args.duration <= 0) {
    throw new Error("--duration must be a positive integer");
  }
  if (!Number.isInteger(args.routeConcurrency) || args.routeConcurrency <= 0) {
    throw new Error("--route-concurrency must be a positive integer");
  }
  return args;
}

function freePort() {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.once("error", reject);
    server.listen(0, HOST, () => {
      const { port } = server.address();
      server.close(() => resolve(port));
    });
  });
}

function tryConnect(port) {
  return new Promise((resolve) => {
    const socket = net.createConnection({ host: HOST, port }, () => {
      socket.destroy();
      resolve(true);
    });
    socket.once("error", () => {
      socket.destroy();
      resolve(false);
    });
    socket.setTimeout(200, () => {
      socket.destroy();
      resolve(false);
    });
  });
}

async function waitForPort(port, timeoutMs = 15000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (await tryConnect(port)) return;
    await new Promise((r) => setTimeout(r, 20));
  }
  throw new Error(`server on ${HOST}:${port} did not become ready`);
}

function writeTemp(prefix, name, contents) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), prefix));
  const file = path.join(dir, name);
  fs.writeFileSync(file, contents, "utf8");
  return file;
}

function wrkLua(sendKind) {
  const lines = [
    'wrk.method = "POST"',
    `wrk.body = '${BODY}'`,
    'wrk.headers["Content-Type"] = "application/json"',
  ];
  if (sendKind) lines.push(`wrk.headers["kind"] = "${KIND}"`);
  return lines.join("\n") + "\n";
}

function runWrk(port, connections, duration, luaPath) {
  const threads = Math.min(connections, os.cpus().length || 8);
  const result = spawnSync(
    "wrk",
    [
      `-t${threads}`,
      `-c${connections}`,
      `-d${duration}s`,
      "-s",
      luaPath,
      `http://${HOST}:${port}${PATH}`,
    ],
    { encoding: "utf8" },
  );
  if (result.error) {
    throw new Error(`failed to run wrk: ${result.error.message}`);
  }
  if (result.status !== 0) {
    throw new Error(
      `wrk exited with status ${result.status}: ${(result.stderr || "").trim()}`,
    );
  }
  const out = result.stdout || "";
  const req = REQ_RE.exec(out);
  const lat = LAT_RE.exec(out);
  return { rps: req ? Number.parseFloat(req[1]) : 0, lat: lat ? lat[1] : "?" };
}

// --- server launchers --------------------------------------------------------
// Each returns { port, sendKind, proc }. The caller awaits readiness and kills
// `proc` when done.

async function spawnNode(source, prefix, port) {
  const script = writeTemp(prefix, "server.js", source);
  const proc = spawn(process.execPath, [script], {
    stdio: ["ignore", "ignore", "pipe"],
  });
  let stderr = "";
  proc.stderr.on("data", (chunk) => {
    stderr += chunk.toString();
  });
  let ready = false;
  const exitPromise = new Promise((_, reject) => {
    proc.once("exit", (code, signal) => {
      // Once ready, exit is the expected teardown; only treat it as a crash
      // while we are still waiting for the port.
      if (ready) return;
      reject(
        new Error(
          `server (${prefix}) exited before becoming ready ` +
            `(code=${code}, signal=${signal}):\n${stderr.trim()}`,
        ),
      );
    });
  });
  await Promise.race([waitForPort(port), exitPromise]);
  ready = true;
  return proc;
}

async function mqbServer(routeConcurrency, variant = "kind") {
  const port = await freePort();
  const config = `
routes:
  http_bench:
    concurrency: ${routeConcurrency}
    batch_size: 128
    input:
      http:
        url: "${HOST}:${port}"
        path: "${PATH}"
        method: "POST"
        internal_buffer_size: 8192
        request_timeout_ms: 30000
        concurrency_limit: 512
        inline_response_fast_path: true
    output:
      response: {}
`;
  // "kind": addHandler delivers a JS object decoded by Rust/napi and returns
  //   Message.fromJson (napi marshals the object back to serde_json). Two JSON
  //   round-trips cross the boundary as structured values.
  // "raw": withHandler hands the body Buffer straight to JS; V8 does the
  //   JSON.parse/stringify and we return raw bytes — no napi value marshaling.
  const handler =
    variant === "raw"
      ? `route.withHandler((m) => {
           const d = JSON.parse(m.payload);
           d.value += 1;
           return new Message(Buffer.from(JSON.stringify(d)));
         });`
      : `route.addHandler(${JSON.stringify(KIND)}, (data) => Message.fromJson({ value: data.value + 1 }));`;
  // route.start() deploys on a background runtime; the handler runs on THIS
  // process's event loop via the threadsafe function, so we must keep the loop
  // alive without blocking it (no join()) or the callbacks would never run.
  const source = `
"use strict";
const { Route, Message } = require(${JSON.stringify(MQB_MAIN)});
const route = Route.fromStr(${JSON.stringify(config)}, "http_bench");
${handler}
route.start();
const keepAlive = setInterval(() => {}, 1 << 30);
process.on("SIGTERM", () => { try { route.stop(); } catch (_) {} clearInterval(keepAlive); process.exit(0); });
`;
  const proc = await spawnNode(source, "mqb-node-", port);
  return { port, sendKind: variant === "kind", proc };
}

async function nodeHttpServer() {
  const port = await freePort();
  const source = `
"use strict";
const http = require("node:http");
const server = http.createServer((req, res) => {
  if (req.method !== "POST") { res.writeHead(404); res.end(); return; }
  let body = "";
  req.on("data", (c) => { body += c; });
  req.on("end", () => {
    const data = JSON.parse(body);
    data.value += 1;
    const rb = JSON.stringify(data);
    res.writeHead(200, { "content-type": "application/json" });
    res.end(rb);
  });
});
server.listen(${port}, ${JSON.stringify(HOST)});
`;
  const proc = await spawnNode(source, "node-http-", port);
  return { port, sendKind: false, proc };
}

async function fastifyServer() {
  const port = await freePort();
  const source = `
"use strict";
const Fastify = require(${JSON.stringify(resolvePeer("fastify"))});
const app = Fastify({ logger: false });
app.post(${JSON.stringify(PATH)}, async (req) => { const d = req.body; d.value += 1; return d; });
app.listen({ port: ${port}, host: ${JSON.stringify(HOST)} });
`;
  const proc = await spawnNode(source, "fastify-", port);
  return { port, sendKind: false, proc };
}

async function uwsServer() {
  const port = await freePort();
  // uWebSockets.js is the Node HTTP leader on public benchmarks (http-arena,
  // TechEmpower): a C++ socket layer that bypasses Node's `http` stack.
  const source = `
"use strict";
const uWS = require(${JSON.stringify(resolvePeer("uWebSockets.js"))});
uWS.App()
  .post(${JSON.stringify(PATH)}, (res, req) => {
    res.onAborted(() => {});
    const chunks = [];
    res.onData((ab, isLast) => {
      chunks.push(Buffer.from(ab));
      if (isLast) {
        const data = JSON.parse(Buffer.concat(chunks));
        data.value += 1;
        const rb = JSON.stringify(data);
        res.cork(() => {
          res.writeHeader("Content-Type", "application/json").end(rb);
        });
      }
    });
  })
  .listen(${JSON.stringify(HOST)}, ${port}, (token) => { if (!token) process.exit(1); });
`;
  const proc = await spawnNode(source, "uws-", port);
  return { port, sendKind: false, proc };
}

async function expressServer() {
  const port = await freePort();
  const source = `
"use strict";
const express = require(${JSON.stringify(resolvePeer("express"))});
const app = express();
app.use(express.json());
app.post(${JSON.stringify(PATH)}, (req, res) => { const d = req.body; d.value += 1; res.json(d); });
app.listen(${port}, ${JSON.stringify(HOST)});
`;
  const proc = await spawnNode(source, "express-", port);
  return { port, sendKind: false, proc };
}

function resolvePeer(name) {
  // Resolve relative to this package so children find the dep no matter their cwd.
  return require.resolve(name, { paths: [path.join(__dirname, "..")] });
}

function peerInstalled(name) {
  try {
    resolvePeer(name);
    return true;
  } catch (_) {
    return false;
  }
}

const TARGETS = {
  mqb: { run: (a) => mqbServer(a.routeConcurrency, "kind"), needs: null },
  "mqb-raw": { run: (a) => mqbServer(a.routeConcurrency, "raw"), needs: null },
  "node-http": { run: () => nodeHttpServer(), needs: null },
  uws: { run: () => uwsServer(), needs: "uWebSockets.js" },
  fastify: { run: () => fastifyServer(), needs: "fastify" },
  express: { run: () => expressServer(), needs: "express" },
};

async function killProc(proc) {
  if (!proc || proc.exitCode !== null) return;
  proc.kill("SIGTERM");
  await new Promise((resolve) => {
    const timer = setTimeout(() => {
      proc.kill("SIGKILL");
      resolve();
    }, 5000);
    proc.once("exit", () => {
      clearTimeout(timer);
      resolve();
    });
  });
}

async function main() {
  if (spawnSync("wrk", ["--version"], { encoding: "utf8" }).error) {
    throw new Error("wrk not found on PATH (brew install wrk)");
  }
  const args = parseArgs(process.argv.slice(2));
  const requested = args.targets
    .split(",")
    .map((t) => t.trim())
    .filter(Boolean);

  const targets = [];
  for (const t of requested) {
    const target = TARGETS[t];
    if (!target) throw new Error(`unknown target '${t}'; choices: ${Object.keys(TARGETS).join(", ")}`);
    if (target.needs && !peerInstalled(target.needs)) {
      console.log(`skipping ${t}: '${target.needs}' not installed (npm i -D ${target.needs})`);
      continue;
    }
    targets.push(t);
  }
  if (!targets.length) throw new Error("no runnable targets");

  const results = {};
  for (const target of targets) {
    results[target] = {};
    let server;
    try {
      server = await TARGETS[target].run(args);
    } catch (error) {
      console.log(`skipping ${target}: ${error.message}`);
      continue;
    }
    try {
      const lua = writeTemp("wrk-", "post.lua", wrkLua(server.sendKind));
      runWrk(server.port, Math.max(...args.connections), 2, lua); // warmup
      for (const conn of args.connections) {
        const { rps, lat } = runWrk(server.port, conn, args.duration, lua);
        results[target][conn] = rps;
        console.log(
          `${target.padEnd(12)} c=${String(conn).padEnd(4)} ${rps
            .toLocaleString("en-US", { maximumFractionDigits: 0 })
            .padStart(12)} req/s  (lat ${lat})`,
        );
      }
    } finally {
      await killProc(server.proc);
    }
  }

  console.log("\nreq/s by connections (native wrk load):");
  const header = "target".padEnd(12) + args.connections.map((c) => String(c).padStart(14)).join("");
  console.log(header);
  for (const target of targets) {
    const row =
      target.padEnd(12) +
      args.connections
        .map((c) =>
          (results[target][c] || 0)
            .toLocaleString("en-US", { maximumFractionDigits: 0 })
            .padStart(14),
        )
        .join("");
    console.log(row);
  }
}

main().catch((error) => {
  console.error(error.message || error);
  process.exitCode = 1;
});
